WebApp/recommendations.py

```diff
@@ -0,0 +1,164 @@
+import os
+import sys
+import random
+import math
+import numpy as np
+import pandas as pd
+import scipy
+
+def get_books_from_indices(top_book_indices, books):
+    top_books = []
+    for i in range(len(top_book_indices)):
+        book_id = top_book_indices[i]
+        book = books.iloc[book_id - 1]  # index is book_id - 1
+        book['rank'] = i + 1
+
+        # for some reason, some of the text fields have newlines appended to them
+        book['title'] = book['title'].strip()
+        book['author'] = book['author'].strip()
+        top_books.append(book)
+    return top_books
+
+
+def get_top_n_recs(result, books, n, q):
+    recs = []
+    for i in range(len(result)):
+        if q[i] == 0:  # book user hasn't already rated
+            recs.append((i, result[i]))
+        else:
+            recs.append((i, float('-inf')))
+    recs = sorted(recs, key=lambda tup: tup[1], reverse=True)
+    print("finished sorting recs")
+
+    top_books = []
+    for i in range(n):
+        book_id = recs[i][0]
+        book = books.iloc[book_id]
+        book['rank'] = i + 1
+
+        # for some reason, some of the text fields have newlines appended to them
+        book['title'] = book['title'].strip()
+        book['author'] = book['author'].strip()
+        top_books.append(book)
+
+    return top_books
+
+
+def most_popular(books, n):
+    top_books = []
+    for i in range(n):
+        book = books.iloc[i]
+        book['rank'] = i + 1
+
+        # for some reason, some of the text fields have newlines appended to them
+        book['title'] = book['title'].strip()
+        book['author'] = book['author'].strip()
+        top_books.append(book)
+
+    return top_books
+
+
+def map_user(q, V):
+    # map new user to concept space by q*V
+    user_to_concept = np.matmul(q, V)
+    # map user back to itme space with user_to_concept * VT
+    result = np.matmul(user_to_concept, V.T)
+    return result
+
+def map_user_to_features(p, features):
+    p_sparse = scipy.sparse.csr_matrix(p)
+    # map new user to concept space by p*features
+    user_to_concept = p_sparse.dot(features)
+    # map user back to itme space with user_to_concept * featuresT
+    result = user_to_concept.dot(features.T).todense()
+    return result.T
+
+def get_predictions(p, q, user_bias, item_bias, global_bias):
+    pred_ratings = np.zeros(len(q))
+    for i in range(len(q)):
+        pred = global_bias + user_bias + item_bias[i] + np.dot(p, q[i])
+        pred_ratings[i] = pred
+    return pred_ratings
+
+def partial_fit(new_user_ratings, q, item_bias, global_bias):
+
+    # create array of indices of books this user has actually rated
+    indices = []
+    for i in range(len(new_user_ratings)):
+        if new_user_ratings[i] != 0:
+            indices.append(i)
+
+    # Hyperparams
+    learning_rate = 0.07
```

Comment:
We should move all this hardcoded stuff somewhere else, but idek where, and not right now.